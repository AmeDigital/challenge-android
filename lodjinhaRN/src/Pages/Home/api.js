export const getCategories = async () => {
  const response = await fetch('https://alodjinha.herokuapp.com/categoria');
  const categories = await response.json();
  return categories
}

export const getBestSells = async () => {
  const response = await fetch('https://alodjinha.herokuapp.com/produto/maisvendidos');
  const bestSells = await response.json();
  return bestSells
}

export const getBanners = async () => {
  const response = await fetch('https://alodjinha.herokuapp.com/banner');
  const banners = await response.json();
  return banners
}