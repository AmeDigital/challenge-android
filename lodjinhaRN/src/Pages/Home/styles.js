import { StyleSheet } from 'react-native';

export const Styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FCFCFC'
  },
  categoriesContainer: {
    flex: 1,
    borderBottomWidth: 1,
    borderBottomColor: '#d5d5d5'
  },
  bannerContainer: {
   flex: 1
  },
  bestSellersContainer: {
    flex: 3
  },
  slideImage:{
   height: 120
  },
  titleContainer: {
    borderBottomWidth: 1,
    borderBottomColor: '#d5d5d5',
    padding: 10
  },
  title: {
    fontWeight: 'bold',
    fontSize: 16,
    color: '#2d3142'
  },
  categoryItemContainer: {
    flex: 2,
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 8,
    paddingTop: 8
  },
  categoryItemImage: {
    height: 50,
    width: 50
  },
  categoryItemDescription: {
    
  }
})