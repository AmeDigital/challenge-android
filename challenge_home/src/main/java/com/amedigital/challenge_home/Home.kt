package com.amedigital.challenge_home

import android.util.Log
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.State
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import com.amedigital.challenge_model.Banner
import com.amedigital.challenge_model.Categoria
import com.amedigital.challenge_model.Produto
import com.amedigital.challenge_model.api.Resource
import com.amedigital.challenge_model.fakeCategorias
import com.amedigital.challenge_model.fakeProdutos
import com.amedigital.challenge_produto.CategoriaActivity
import com.amedigital.challenge_produto.ListaCategorias
import com.amedigital.challenge_produto.ListaProdutos
import com.amedigital.challenge_produto.ProdutoActivity
import com.amedigital.challenge_produto.widgets.LogAndShowErrorPanel
import com.amedigital.coreui.widgets.Banner
import com.amedigital.coreui.widgets.WaitingIndicator
import org.koin.androidx.compose.getViewModel

@Composable
fun Home() {
    val viewModel = getViewModel<HomeViewModel>()
    val bannerState = viewModel.banners.observeAsState()
    val categoriaState = viewModel.categorias.observeAsState()
    val maisVendidosState = viewModel.maisVendidos.observeAsState()

    when (val banner = bannerState.value) {
        is Resource.Requesting -> WaitingIndicator()
        is Resource.Failure -> LogAndShowErrorPanel(banner.throwable)
        is Resource.Success -> HomeView(banner.value, categoriaState.value, maisVendidosState.value)
    }
}

@Composable
private fun HomeView(
    images: List<Banner>,
    categorias: Resource<List<Categoria>>?,
    maisVendidos: Resource<List<Produto>>?
) {
    val scrollState = rememberScrollState()
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(scrollState)
    ) {
        Banner(images.map { it.linkUrl }, onImageClick = { image ->
            Log.d("HOME", image)
        })

        when (categorias) {
            is Resource.Requesting -> WaitingIndicator()
            is Resource.Failure -> LogAndShowErrorPanel(categorias.throwable)
            is Resource.Success -> CategoriasView(categorias.value)
        }

        when (maisVendidos) {
            is Resource.Requesting -> WaitingIndicator()
            is Resource.Failure -> LogAndShowErrorPanel(maisVendidos.throwable)
            is Resource.Success -> MaisVendidosView(maisVendidos.value)
        }
    }
}

@Composable
private fun CategoriasView(categorias: List<Categoria>) {
    val context = LocalContext.current
    ListaCategorias(categorias, onCategoriaClick = { categoria ->
        CategoriaActivity.gotoCategoria(context, categoria)
    })
}

@Composable
private fun MaisVendidosView(produtos: List<Produto>) {
    val context = LocalContext.current
    ListaProdutos(produtos, onProdutoClick = { produto ->
        ProdutoActivity.gotoProduto(context, produto)
    })
}