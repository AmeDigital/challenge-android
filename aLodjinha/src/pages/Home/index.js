import React from 'react';
import { View, ScrollView } from 'react-native';

import styles from './styles';

import LogoTitle from '../../components/LogoTitle';
import Banners from '../../components/Banners';
import Categories from '../../components/Categories';
import BestSellers from '../../components/BestSellers';

function HomeScreen({ navigation }) {
	return (
		<ScrollView>
			<View style={styles.container}>
				<Banners />
				<Categories navigation={navigation} />
				<BestSellers navigation={navigation} />
			</View>
		</ScrollView>
	);
}

HomeScreen.navigationOptions = ({ navigation }) => {
	return {
		headerTitle: <LogoTitle navigation={navigation} />
	};
};

export default HomeScreen;
