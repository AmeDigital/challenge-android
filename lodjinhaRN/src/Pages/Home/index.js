import React, { Component, Fragment } from 'react';
import Swiper from 'react-native-swiper';

import { 
  View, 
  FlatList, 
  Text, 
  Image,
  SafeAreaView, 
  StatusBar,
  Linking,
  TouchableOpacity
} from 'react-native';

import { getBanners, getBestSells, getCategories } from './api';

import ProductItem from '../../components/ProductItem';
import Header from './components/Header';

import { Styles } from './styles';

export default class Home extends Component {

  static navigationOptions = Header

  state = {
    categories: [],
    bestSells: [],
    banners: []
  }

  async componentDidMount() {
      const bestSells = await getBestSells();
      const categories = await getCategories();
      const banners = await getBanners();

      this.setState(
        {
          categories: categories.data,
          bestSells: bestSells.data,
          banners: banners.data
        }
      )
  }

  _keyExtractor = (item, index) => String(item.id);

  handleClick = (url) => {
    Linking.canOpenURL(url).then(supported => {
      if (supported) {
        Linking.openURL(url);
      }
    });
  };

  renderCategoryItem = ({ item }) => (
    <TouchableOpacity 
      style={Styles.categoryItemContainer} 
      onPress={() => this.props.navigation.navigate('Category', { category: item })}>
      <Image style={Styles.categoryItemImage} source={{ uri: item.urlImagem }} />
      <Text style={Styles.categoryItemDescription}>{item.descricao}</Text>
    </TouchableOpacity>
  )

  render() {
    return (
      <Fragment>
        <StatusBar barStyle="dark-content" />
          <SafeAreaView style={Styles.container}>
              <Swiper autoplay={true} style={Styles.bannerContainer}>
                {
                  this.state.banners.map((item) => (
                    <TouchableOpacity key={item.id} onPress={() => this.handleClick(item.linkUrl)}>
                      <Image 
                        style={Styles.slideImage} 
                        resizeMode="cover"
                        source={{ uri: item.urlImagem }} />
                    </TouchableOpacity>
                  ))
                }
              </Swiper>
              <View style={Styles.categoriesContainer}>
                <View style={Styles.titleContainer}>
                  <Text style={Styles.title}>Categorias</Text>
                </View>
                <FlatList
                  data={this.state.categories}
                  horizontal
                  showsHorizontalScrollIndicator={false}
                  keyExtractor={this._keyExtractor}
                  renderItem={this.renderCategoryItem} />
              </View>
              <View style={Styles.bestSellersContainer}>
                <View style={Styles.titleContainer}>
                  <Text style={Styles.title}>Mais vendidos</Text>
                </View>
                <FlatList
                    data={this.state.bestSells}
                    keyExtractor={this._keyExtractor}
                    renderItem={({ item }) => <ProductItem item={item} />} />
              </View>
          </SafeAreaView>
      </Fragment>
    );
  }
}
