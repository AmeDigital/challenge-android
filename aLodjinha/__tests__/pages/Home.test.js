import React from 'react';

import renderer from 'react-test-renderer';

import Home from '../../src/pages/Home';

it('renders correctly', () => {
	const tree = renderer.create(<Home />).toJSON();
	expect(tree).toMatchSnapshot();
});
