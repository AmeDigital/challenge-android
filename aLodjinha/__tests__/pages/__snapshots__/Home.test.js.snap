// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#ccc",
            "elevation": 2,
            "height": 150,
            "justifyContent": "center",
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <ActivityIndicator
            animating={true}
            color="#5e2a84"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "borderBottomColor": "#aaaaaa",
            "borderBottomWidth": 1,
            "height": 200,
          }
        }
      >
        <View
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#aaaaaa",
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2d3142",
                "fontSize": 35,
                "fontWeight": "bold",
              }
            }
          >
            Categorias
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <ActivityIndicator
            animating={true}
            color="#5e2a84"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#aaaaaa",
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2d3142",
                "fontSize": 35,
                "fontWeight": "bold",
              }
            }
          >
            Mais Vendidos
          </Text>
        </View>
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <ActivityIndicator
              animating={true}
              color="#5e2a84"
              hidesWhenStopped={true}
              size="large"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
